### PR TITLE
feat: add `flash_attention 3` kernel for diffusers pipelines

### PR DESCRIPTION
**Pull request by johannaSommer**
## Description
In this PR, we make use of HuggingFace's kernel hub to utilize flash attention 3 in diffusers pipelines where possible. Importantly, diffusers is refactoring their attention handling right now, hence we have two different implementations. 

In the current diffusers version, we wrap the call of a pipeline to intercept each call to `torch.nn.funcional.scaled_dot_product_attention`. If shapes, dtype and keyword arguments can be supported by flash attention, we reroute the computation to the fa3 kernel.

In the newer diffusers version (0.35.0, unreleased so far), we can register a new attention backend that calls the native attention or fa3 on the same criteria.

Overall, the speedup is minor for T2I pipelines but is substantial for Video Gen pipelines, in particular Wan.

## Related Issue
None.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Added algorithm tests with Flux and Wan and added combination tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Additional Notes
Flash Attention 3 on WAN gives around a 1.4 speedup, no warmup or quality degradation of course.

Minimal Script:
``` python
from pruna import SmashConfig, smash
from diffusers import WanPipeline
import torch

pipeline = WanPipeline.from_pretrained(
    "Wan-AI/Wan2.2-T2V-A14B-Diffusers", torch_dtype=torch.bfloat16
)
pipeline.to("cuda");

config = SmashConfig()
config._prepare_saving = False
config["kernel"] = "flash_attn3"
pipeline = smash(pipeline, config)

prompt = "A cat is doing an acrobatic dive into a swimming pool at the olympics, from a 10m high diving board, flips and spins"
negative_prompt = "Bright tones, overexposed, static, blurred details, subtitles, style, works, paintings, images, static, overall gray, worst quality, low quality, JPEG compression residue, ugly, incomplete, extra fingers, poorly drawn hands, poorly drawn faces, deformed, disfigured, misshapen limbs, fused fingers, still picture, messy background, three legs, many people in the background, walking backwards"
output = pipeline.__call__(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=480,
    width=640,
    num_frames=69,
    guidance_scale=5.0,
    num_inference_steps=10,
    generator=torch.Generator(device="cpu").manual_seed(1),
).frames[0]
```
Output during Generation:
`100%|██████████| 10/10 [01:04<00:00,  6.45s/it]`

----
Generation times on the prompt / size above on 1 H100:
- base: 8.04 s/it
- `flash_attn3`: 6.46 s/it (no warmup!)
- `torch.compile`: 6.38 s/it
- `flash_attn3` + `torch.compile`: 5.15 s/it

### CODE REVIEW

**Comment by johnrachwan123** [comment]
bugbot run

File: src/pruna/algorithms/kernels/flash_attn3.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+        ----------
+        model : Any
+            The model to check.
+
+        Returns
+        -------
+        bool
+            True if the model is a valid model for the algorithm, False otherwise.
+        """
+        if Version(diffusers_version) >= Version("0.35.0.dev0"):
```
Comment (sharpenb) [inline src/pruna/algorithms/kernels/flash_attn3.py:67]: I am not sure about these conditions. It might be able to work with transformers.

File: src/pruna/algorithms/kernels/flash_attn3.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+            # fa3 only supports attention with num_query_heads % num_kv_heads == 0
+            num_heads_pass = all(query.shape[1] % t.shape[1] == 0 for t in (key, value))
+
+            # test head dimension
+            head_dim_pass = all(t.shape[3] <= 256 for t in (query, key, value))
+
+            # if any constraints are not met or unsupported input arguments are being used, reroute to native attention
+            if attn_mask is not None or dropout_p != 0.0 or not dtype_pass or not num_heads_pass or not head_dim_pass:
+                pruna_logger.debug(
```
Comment (sharpenb) [inline src/pruna/algorithms/kernels/flash_attn3.py:208]: This debugging message does not really say what are the constraints which are explained in the comments above.

File: src/pruna/algorithms/kernels/flash_attn3.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+        Returns
+        -------
+        Any
+            The wrapped model.
+        """
+        imported_packages = self.import_algorithm_packages()
+
+        # in the new version of diffusers, we can use the modular attention backend to inject flash_attn3
+        if Version(diffusers_version) >= Version("0.35.0.dev0"):
```
Comment (sharpenb) [inline src/pruna/algorithms/kernels/flash_attn3.py:100]: Do we need to support both versions for certain model use-cases? It feels to me that the code could be more readable by saying we only support from version xyz.

File: src/pruna/algorithms/kernels/flash_attn3.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+            # replace in all compatible components
+            for component in model.components.values():
+                if hasattr(component, "set_attention_backend") and component.dtype in [
+                    torch.bfloat16,
+                    torch.float16,
+                ]:
+                    component.set_attention_backend("flash_attn3_pruna")
+
+        else:
```
Comment (sharpenb) [inline src/pruna/algorithms/kernels/flash_attn3.py:112]: I was confused by some functions redundant but I guess they are designed for the two diffusers versions. Could we mark e.g. in their docstrings, or separate files.

File: src/pruna/algorithms/kernels/flash_attn3.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    def import_algorithm_packages(self) -> Dict[str, Any]:
+        """
+        Import the algorithm packages.
+
+        Returns
+        -------
+        Dict[str, Any]
+            The algorithm packages.
+        """
+        flash_attention_3 = get_kernel("kernels-community/flash-attn3")
```
Comment (sharpenb) [inline src/pruna/algorithms/kernels/flash_attn3.py:126]: This is designed only for fa3 but `kernels_hub` has many more kernels. I feel that extending it to all kernels directly would be nice. Probably not for this PR but worth spending a bit of time now if the logic is similar.

**Review by sharpenb — CHANGES_REQUESTED** [review]
Thanks for the PR :)  
- I would be interested in a small benchmark (base, torch compile, fa3, fa3+torch.compile) for Wan (if that is the main model targeted by this update) 
- Was it tested with EvaluationAgent after compression?
- Could we have test notebook since it integrates a new compression algorithm?

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 1 above)
Comment (johannaSommer) [reply src/pruna/algorithms/kernels/flash_attn3.py:67]: I don't understand this comment, can you elaborate? Because of refactoring in the upcoming diffusers version we have to apply seperate logic.

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 5 above)
Comment (johannaSommer) [reply src/pruna/algorithms/kernels/flash_attn3.py:126]: I don't this this makes sense here - at least currently the other kernels available on kernel hub tackle other functionality, i.e. not the attention function. I think it is better to think of this `get_kernel` like an aggregation of building, registering and importing a certain kernel package, but which functions you use and how you use them will be different from kernel to kernel. If there are other interesting attention kernels coming soon, we will of course consolidate them.

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 2 above)
Comment (johannaSommer) [reply src/pruna/algorithms/kernels/flash_attn3.py:208]: I know what you mean - I added a slight extension to point to the file where the user could review these criteria. Note that this is really only for debugging purposes, activating this debug message will print _a lot_ since the text is displayed at **every** `torch.nn.functional.scaled_dot_product_attention` call. I added this mainly so that we can support someone who tries this on a pipeline we haven't tested before and doesn't experience any speedup.

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 1 above)
Comment (sharpenb) [reply src/pruna/algorithms/kernels/flash_attn3.py:67]: Sorry for the confusion. I mean that model checks focus on diffusers, while it might work for transformers architectures. My understanding now is that it was not tested for transformers but could be worth to have a quick check.

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 5 above)
Comment (sharpenb) [reply src/pruna/algorithms/kernels/flash_attn3.py:126]: I think that it aligns with my point. Here, would it be possible to have a streamlined way to integrate other kernels e.g. by defining a mapping from kernel to the functions we use. I would expect that if we implement another kernel from the hub, we would go over similar implementation steps e.g. registering, identifying functions, and wrapping calls. One potential ideais to take inspiration from their blog which suggests to use `use_kernel_forward_from_hub`. Again no need to refactorize , but wanted to highlight this since it might be relevant for the future ;)

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 2 above)
Comment (sharpenb) [reply src/pruna/algorithms/kernels/flash_attn3.py:208]: Super nice thanks!

File: src/pruna/algorithms/kernels/flash_attn3.py (hunk 3 above)
Comment (sharpenb) [reply src/pruna/algorithms/kernels/flash_attn3.py:100]: Let's have a reminder to remove support for older version of diffusers when a couple fo diffusers releases will have been done :)

**Review by sharpenb — APPROVED** [review]
Let's go! I left some final thoughts
